bluetooth: controller: Ext. adv. duration handling fix

Tronil opened the pull request:
- Remaining duration is now in us instead of ticks to avoid overflow when ticks have higher resolution
- Account for the added random delay when updating remaining duration
- Re-worked JIT re-scheduling update a bit to keep the random delay information and remaining duration correct
- Add EVENT_OVERHEAD_START_US to ticks_anchor when enabling advertising, so the first advertising event is not late due to calculations for aux packets etc.

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
@cvinayak @mtpr-ot here is the proposed changes to the handling of the extended advertising duration limit as discussed in the last controller meeting. Feedback welcome :-)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1995,11 +2004,11 @@ void ull_adv_done(struct node_rx_event_done *done)
-		    HAL_TICKER_US_TO_TICKS((uint64_t)adv->interval *
-			ADV_INT_UNIT_US))) {
-		adv->ticks_remain_duration = 0;
+	} else if (adv->remain_duration_us &&
+		   (adv->remain_duration_us <=
+		    (adv->interval * ADV_INT_UNIT_US + EVENT_OVERHEAD_START_US +
+		     HAL_TICKER_TICKS_TO_US(adv->delay)))) {
+		adv->remain_duration_us = 0;
```
as `adv->remain_duration_us = interval_us;`

```suggestion
	} else if (adv->remain_duration_us &&
		   (adv->remain_duration_us <=
		    (adv->interval * ADV_INT_UNIT_US +
		     auto_variable_for_adv_delay_of_JIT_scheduling))) {
		adv->remain_duration_us = 0;
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1908,7 +1908,7 @@ static uint32_t ticker_update_rand(struct ll_adv_set *adv, uint32_t ticks_delay_
 		  (fp_op_func == NULL));
 
-#if defined(CONFIG_BT_CTLR_JIT_SCHEDULING)
+#if defined(CONFIG_BT_CTLR_JIT_SCHEDULING) || defined(CONFIG_BT_CTLR_ADV_EXT)
```
```suggestion
#if defined(CONFIG_BT_CTLR_JIT_SCHEDULING) && defined(CONFIG_BT_CTLR_ADV_EXT)
```

cvinayak reviewed (changes requested):
1. Please have a commit that converts the non-JIT scheduling code to 10 ms units (duration in HCI is 10 ms units), this needs less storage bits (16-bits).
2. Next commit change the JIT scheduling to use the 10 ms unit.
3. `adv->delay` is not needed by non-JIT scheduling implementation.

Tronil commented:
> 1. Please have a commit that converts the non-JIT scheduling code to 10 ms units (duration in HCI is 10 ms units), this needs less storage bits (16-bits).
>     2. Next commit change the JIT scheduling to use the 10 ms unit.

Using 10 ms units would introduce too much error the way we handle the duration (ie. subtracting from it after every advertising event). Unless you were thinking of using a starting timestamp instead of subtracting - but then you will to use more storage in the end to store the start timestamp.

>     3. `adv->delay` is not needed by non-JIT scheduling implementation.

It is used to account for the random delay added to each advertising event, so it is needed for non-JIT scheduling as well (unless I am missing something regarding the ticker implementation).

cvinayak commented:
> > 1. Please have a commit that converts the non-JIT scheduling code to 10 ms units (duration in HCI is 10 ms units), this needs less storage bits (16-bits).
> >    2. Next commit change the JIT scheduling to use the 10 ms unit.
> 
> Using 10 ms units would introduce too much error the way we handle the duration (ie. subtracting from it after every advertising event). Unless you were thinking of using a starting timestamp instead of subtracting - but then you will to use more storage in the end to store the start timestamp.

Updated:
1. Have a commit that converts the non-JIT scheduling code to 1 us unit
2. Next commit change the JIT scheduling to use the 1 us unit.

Tronil committed:
bluetooth: controller: Ext. adv. duration handling fix

- Remaining duration is now in us instead of ticks to avoid
  overflow when ticks have higher resolution
- First advertising event should not reduce remain duration
- Add EVENT_OVERHEAD_START_US to ticks_anchor when enabling
  advertising, so the first advertising event is not late due to
  calculations for aux packets etc.
- Update some bsim tests to account for timing difference
  introduced by above change

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil committed:
bluetooth: controller: JIT fixes for ext. adv. duration handling

- Account for the added random delay when updating remaining duration
- Re-worked JIT re-scheduling update a bit to keep the random delay
  information and remaining duration correct

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
Reworked the implementation to make use of ticks_drift instead of adv->delay when possible.

>     1. Have a commit that converts the non-JIT scheduling code to 1 us unit
> 
>     2. Next commit change the JIT scheduling to use the 1 us unit.

I have split into two commits - one part for the non-JIT related changes and another for the changes that are only related to the JIT scheduler.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -2290,21 +2301,25 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t ticks_drift,
-
-			if (adv->ticks_remain_duration > ticks_elapsed) {
-				adv->ticks_remain_duration -= ticks_elapsed;
+		if (adv->remain_duration_us && adv->event_counter > 0) {
```
Why `adv->event_counter > 0` ?

If say, first 2 events where skipped due to overlap, then `adv->remain_duration_us` would not be reduced as we will not enter this if-clause.

```suggestion
		adv->event_counter += (lazy + 1);
		
		if (adv->remain_duration_us) {
```

Tronil replied:
According to the spec (vol 4, part E, section 7.8.56):
> The Duration[i] parameter indicates the duration for which that advertising set
is enabled. The duration begins at the start of the first advertising event of this
advertising set.

Before, we would actually end the advertising one event too early since we subtracted one interval when the first advertising event was started.

The skipping of events is a bit problematic, yes; I'm not sure what the correct thing to do is there. On one hand you could argue that the first advertising event is the one actually sent out on the radio; On the other hand that would lead to less predictable behavior for the host, so maybe it should be counted as if they were sent.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -2290,21 +2301,25 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t ticks_drift,
+				adv->remain_duration_us = interval_us;
 			}
 		}
+
+		adv->event_counter += (lazy + 1);
```
```suggestion
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -2290,21 +2301,25 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t ticks_drift,
+						 HAL_TICKER_TICKS_TO_US(ticks_drift);
+
+			/* End advertising if the added random delay pushes us beyond the limit */
+			if (adv->remain_duration_us > elapsed_us + interval_us +
+						      HAL_TICKER_TICKS_TO_US(random_delay)) {
```
`elapsed_us` consists of current event's `ticks_drift` plus no. of skipped intervals (represented by `lazy`) before the current event plus the next interval (`+ 1`).

I agree with you for adding `random_delay`  which is guess here in the `ticker_cb`  as the ticks slow window feature would nudge it when there are overlaps, and hence in the next interval's `ticker_cb` we will use the actual `ticks_drift` to reduce the `remain_duration_us`.

```suggestion
			if (adv->remain_duration_us > elapsed_us +
						      HAL_TICKER_TICKS_TO_US(random_delay)) {
```

cvinayak reviewed (dismissed):
The upstream non-JIT scheduling implement has passed conformance tests related to duration and event counts, hence I am being extra cautious with the change in this PR.

cvinayak reviewed (commented):
I have created a PR combining some of my open PR and separated you PR into more commits, I expect to not see failures to mesh or eatt tests there.

Review https://github.com/zephyrproject-rtos/zephyr/pull/51012 and let me know if I missed anything from your PR.

Tronil commented:
Superseded by PR https://github.com/zephyrproject-rtos/zephyr/pull/51012, closing this one
